File: server/public/tools/elements/widgets/GtInput.ts
import {Element, Property, Listener} from "../../polymer/Annotations";
import {GtForm} from "./GtForm";
import {PolymerElement} from "../../polymer/PolymerElement";

/**
 * Text input
 */
@Element({
	selector: "gt-input",
	template: "/assets/imports/widgets.html",
	dependencies: [GtForm]
})
export class GtInput extends PolymerElement {
	/**
	 * Disable the input
	 */
	@Property({ reflect: true })
	public disabled: boolean;

	/**
	 * Prevent form submission if empty
	 */
	@Property({ reflect: true })
	public required: boolean;

	/**
	 * Input type "text", "password"
	 */
	@Property
	public type: string = "text";

	/**
	 * Error message
	 */
	@Property({ reflect: true, notify: true })
	public error: string = null;

	/**
	 * Proxy to input value
	 */
	@Property({ notify: true })
	public value: string = "";

	/**
	 * Focus the input
	 */
	public focus() {
		this.$.input.focus();
	}

	/**
	 * Blur the input
	 */
	public blur() {
		this.$.input.blur();
	}

	/**
	 * Catch Enter-key presses in the input and forward to the form
	 */
	@Listener("input.keypress")
	private InputKeypress(e: KeyboardEvent) {
		if (e.keyCode == 13) {
			e.preventDefault();
			const form = this.host(GtForm);
			if (form) {
				this.blur();
				form.submit();
			}
		}
	}

	@Listener("input.keyup")
	private InputUp(e: KeyboardEvent) {
		this.debounce("value-changed", () => this.value = this.$.input.value, 200);
	}

	public click() {
		this.focus();
	}

	/**
	 * Reflect input focused state to the outer <label> element
	 */
	@Listener("input.focus")
	private InputFocus() {
		this.toggleAttribute("focused", true);
	}

	@Listener("input.blur")
	private InputBlur() {
		this.toggleAttribute("focused", false);
	}
}

File: client/src/gt/component/widget/form/GtInput.scala
package gt.component.widget.form

import gt.component.GtHandler
import org.scalajs.dom.Event
import org.scalajs.dom.raw.{HTMLInputElement, KeyboardEvent}
import scala.concurrent.duration._
import util.Debouncer
import util.implicits._
import util.jsannotation.js
import xuen.Component

object GtInput extends Component[GtInput](
	selector = "gt-input",
	templateUrl = "/assets/imports/widgets.html"
)

@js class GtInput extends GtHandler with AbstractInput {
	/** Disables the input */
	// TODO: make this sane again once Scala.js DOM is fixed
	this.dyn.disabled = attribute[Boolean].dyn

	/** Prevents form submission if empty */
	val required = attribute[Boolean]

	/** Input type */
	val `type` = attribute[String] := "text"

	/** Input label */
	val label = attribute[String]

	/** Proxy to input value */
	val value = model[String] := ""

	/** Error message */
	val error = property[String]

	/** Is an error message present? */
	val hasError = error ~ { e => e != null && !e.trim.isEmpty }

	/** Focuses the input */
	override def focus(): Unit = child.input.focus()

	/** Removes input focus */
	override def blur(): Unit = child.input.blur()

	// Update the cached value of the input
	private val update = Debouncer(200.millis) {
		value := child.as[HTMLInputElement].input.value
	}

	def reset(): Unit = {
		value := ""
		error := null
	}

	def validate(): Boolean = {
		if (required && value.matches("^\\s*$")) {
			error := "This field is required"
			focus()
			false
		} else {
			error := null
			true
		}
	}

	// Track input state and value
	listen("focus", child.input) { e: Event => setAttribute("focused", "") }
	listen("blur", child.input) { e: Event => removeAttribute("focused") }
	listen("keyup", child.input) { e: Event => update.trigger() }
	listen("change", child.input) { e: Event => update.now() }

	// Capture enter key presses
	listen("keypress", child.input) { e: KeyboardEvent =>
		if (e.keyCode == 13) {
			e.preventDefault()
			update.now()
			for (form <- Option(closest("gt-form").asInstanceOf[GtForm])) {
				blur()
				form.submit()
			}
		}
	}
}

